### ISSUE TITLE
systemd-227: ForwardDelaySec

### ISSUE THREAD

**Issue by vovan69** [systemd/systemd#1710]
Hello,

It looks like the ForwardDelaySec sets the bridge parameter in milliseconds instead of seconds:
- in the *.netdev file

[Bridge]
ForwardDelaySec=1
- brctl showstp br0

...
forward delay             0.01                 bridge forward delay       0.01
...

I might be wrong, but anyway it would be nice to double check...

Regards,
Vlad.

**Comment by poettering** [comment]
@ssahani , any idea?

**Comment by ssahani** [comment]
It works similar to ip command.

```
#ip link set br0 type bridge forward_delay 1

#brctl showstp br-test
br-test
 bridge id      8000.82dbf39247fe
 designated root    8000.82dbf39247fe
 root port         0            path cost          0
 max age          20.00         bridge max age        20.00
 hello time        2.00         bridge hello time      2.00
 forward delay         0.01         bridge forward delay       0.01
 ageing time         300.00
 hello timer           0.00         tcn timer          0.00
 topology change timer     0.00         gc timer           0.00
 flags          
```

and

```
 sudo ip link set br-test type bridge forward_delay 100

showstp br-test
br-test
 bridge id      8000.82dbf39247fe
 designated root    8000.82dbf39247fe
 root port         0            path cost          0
 max age          20.00         bridge max age        20.00
 hello time        2.00         bridge hello time      2.00
 forward delay         1.00         bridge forward delay       1.00
 ageing time         300.00
 hello timer           0.00         tcn timer          0.00
 topology change timer     0.00         gc timer           0.00
 flags          
```

**Comment by vovan69** [comment]
True, but the name "ForwardDelaySec" is confusing (the ip command uses just "forward_delay" option without mentioning time unit).

**Comment by ssahani** [comment]
yes forward_delay in kernel is stored in jiffies . We need to convert this usec value to jiffies  while setting., that is  b->forward_delay/10000 not  USEC_PER_SEC

we need one macro to convert from usec_to_jiffies .

**Comment by poettering** [comment]
@ssahani in systemd, we uniformly parse time values in configuration files as seconds if no unit is specified, and if a unit is specified as that. That's widely documented, for example in "systemd.time(5)". This is uniformly done, regardless what the backend actually implements. And we usually (but not always, unfortunately) indicate the assumed unit of seconds, by suffixing the option name with "Sec".

Hence: ForwardDelaySec= is correctly named, and it should parse second values by default, unless a unit is specified. And when writing this in to the kernel, we should convert this if needed to any other unit.

Hence:

```
ForwardDelaySec=500ms       # Should mean 500 milliseconds
ForwardDelaySec=4           # should mean 4 seconds
ForwardDelaySec=77.5        # Should mean 77 seconds plus 500 milliseconds
```

THat's how we do time in systemd, and networkd should do it the same way as the rest, and everything else is a bug...

The function `parse_sec()` will parse time values of this kind, and place the result in an `usec_t`, which is a 64bit unsigned value of microseconds.

**Comment by poettering** [comment]
@ssahani well, yeah, if the kernel in the API uses jiffies, that's completely OK, but we should convert between our usec values to jiffies when invoking those APIs... I presume "ip" has some code for the conversion, but I think it's just about scaling the value with sysconf(_SC_CLK_TCK) or so...

**Comment by poettering** [comment]
Fixed in #1764.
